File: spring-data-r2dbc/src/main/java/com/example/UserService.java
package com.example;

import java.time.LocalDateTime;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@Transactional
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;

    public Flux<User> findAll() {
        return userRepository.findAll();
    }

    public Mono<User> findById(long id) {
        return userRepository.findById(id);
    }

    public Mono<User> create(String name) {
        User user = new User();
        user.setName(name);
        user.setCreatedAt(LocalDateTime.now());
        return userRepository.save(user);
    }

    public Mono<User> update(User user) {
        return userRepository.findById(user.getId())
                             .flatMap(entity -> {
                                 entity.setName(user.getName());
                                 return userRepository.save(entity);
                             });
    }

    public Mono<Void> deleteById(long id) {
        return userRepository.deleteById(id);
    }
}

File: kotlin/database/spring-jdbc/src/main/kotlin/com/example/UserService.kt
package com.example

import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Service
import java.sql.ResultSet

@Service
class UserService(private val jdbcTemplate: JdbcTemplate) {

    fun findById(id: Int): User =
        jdbcTemplate.queryForObject("SELECT id, name FROM user WHERE id=$id")
        { rs, _ -> toUser(rs) } ?: throw NoSuchElementException("id: $id")

    fun findAll(): List<User> =
        jdbcTemplate.query("SELECT * FROM user")
        { rs, _ -> toUser(rs) }

    private fun toUser(rs: ResultSet): User =
        User(rs.getInt("id"), rs.getString("name"))
}
